more commands to use for divergence information in wstool/rosws info

tkruse opened the pull request:
Not to be merged yet, this PR is to allow travis to check for compatibility and to summarize what still needs to be done

TODO:
- [x] git implementation
- [x] git tests
- [x] svn implementation
- [x] svn tests
- [x] hg implementation
- [x] hg tests
- [x] bzr implementation
- [x] bzr tests
- [x] tar implementation + tests (noop functions)
- [x] decide about git fetch (use --fetch)
- [x] decide about hg named branches vs bookmarks (decided: named branches)

- [x] implement ..fetch in vcstools and wstool
- [x] validate works with wstool

- [x] code review (Kentaro), check for leftover print(), TODO, typos, code style
- [x] code review (@wjwwood)

tkruse commented:
this relates to https://github.com/vcstools/wstool/pull/35

tkruse commented:
Related SIG discussion: https://groups.google.com/d/msg/ros-sig-buildsystem/wLINOnzjLpc/g8z-U6OvZNgJ

tkruse commented:
@wkentaro : Other than for Mercurial, I think we are fine. See https://github.com/vcstools/wstool/pull/38 for more work to do.

wkentaro commented:
Okay, thanks.

tkruse commented:
regarding --fetch, we may need some more implementation, because the other VCSes probably also go to their remotes. So if --fetch is not given we should also avoid executing commands for the other VCSes 
For SVN and BZR: ```if fetch is True: return None```.
For git just don't fetch.
For Mercurial, this is a bit more painful, I need to check...

tkruse commented:
@wkentaro : I am implementing --fetch, BTW, so don't do this on your side as well

tkruse commented:
@wkentaro : plz review my latest changes. Where you do not know the VCS, just make sure cde style is okay, no missing asserts, no left print(), and such.
After that, I think this could be good to merge

wkentaro commented on src/vcstools/git.py:
```diff
@@ -334,15 +334,22 @@ def _do_update(self,
+            result += ' < ' + remote_branch
+        return result
+
+    def get_remote_version(self, spec=None, fetch=False):
```
I think spec is unused.

wkentaro commented on src/vcstools/git.py:
```diff
@@ -334,15 +334,22 @@ def _do_update(self,
+    def get_remote_version(self, spec=None, fetch=False):
+        parent_branch = self.get_branch_parent(fetch=fetch)
+        if parent_branch is not None:
+            return self.get_version(spec='origin/'+parent_branch)
```
Is this okay to assume origin? parent_branch can be `remote2/test_branch`.

wkentaro replied:
Or we don't display remote version when the tracking branch is not origin.
You assume this?

wkentaro replied:
See: https://github.com/tkruse/vcstools/pull/6

tkruse commented:
So, @wkentaro , thanks for the review. If you have nothing else to push, I would like to squash and rebase before the next review and merge. OK?

wkentaro commented:
Yes, thank you.
Please go ahead to next step.

tkruse commented:
@wjwwood : I rebased, squashed and force-pushed. Travis should be happy. If you have time, please review. To manually test, this branch of wstool should make use of the new functions https://github.com/vcstools/wstool/pull/38

wjwwood commented:
I'm going to close and reopen, because for some reason your latest commit, b4c8ebe, has not been picked up by travis.

wjwwood commented:
Hmm, it looks like there is a discrepancy between listing the commits by order and date, so travis is testing @wkentaro's commit, d4a401b. All of the instances of travis seem to fail a style test. It should be trivial to fix up. In the mean time I'll do a code review and test it with the pr on wstool.

wjwwood commented on src/vcstools/bzr.py:
```diff
@@ -180,14 +180,6 @@ def update(self, version='', verbose=False, timeout=None):
-        :returns: the current revision number of the repository. Or if
-          spec is provided, the number of a revision specified by some
-          token.
-        """
```
Why remove the doc block? Is it redundant with other versions of the function, out-of-date, documenting a non-public function, something else?

tkruse replied:
I thought it was duplicate of public api. But it has bzr specific info, so I'll put it back

wjwwood replied:
Ok, there looks to be other instances like this in the pr which might not be duplicates too.

wjwwood commented on test/test_svn.py:
```diff
@@ -176,6 +188,32 @@ def test_checkout_specific_version_and_update_short(self):
+    #    url = self.local_url
+    #    client = SvnClient(path=self.local_path)
+    #    client.checkout(url)
+    #    self.assertEqual(client.get_current_version_label(), 'trunk')
```
Should this be left in or maybe it was accidentally left commented out? I'm fine with leaving it commented out, but maybe a comment explaining why it is disabled and/or when it could be reenabled should go with it.

tkruse replied:
first line is a comment, but really this is not a test, just some WIP, so I'll remove it

wjwwood commented:
Code looks good to me, I'll give it a whirl with wstool asap.

tkruse commented:
@wkentaro : Using the tracked remote is controversial, see 
https://groups.google.com/forum/#!topic/ros-sig-buildsystem/wLINOnzjLpc

I split out those changes, so we can move forward using origin, and discuss the usage of the tracked remote in the thread separately.

wkentaro commented:
You mean C flag, don't you?
As I posted at the google group, I think C flag should mean difference between current branch and tracking branch of specified branch. 
The specified branch must be tracking origin (if not wstool info can display warning) so there is no need to change other functions to track not origin remote except for `vcstools.get_branch_parent()`.

tkruse commented:
@wjwwood : I changed the PR according to code review and fixed the minor pep8 issues, made no other changes. I removed the controversial bit and made a separate PR #99. So this should be good to merge when tests are through, IMO.

wjwwood commented:
Seems to work in my somewhat limited testing. @wkentaro is your review still pending or are you ok with merging this now?

wkentaro commented:
When it comes to https://github.com/vcstools/vcstools/pull/99 I've done the review, and please start working to merge it.
